contracts/ClaimManager.sol

```diff
@@ -0,0 +1,591 @@
+
+  /**
+   * @dev Super-Validator votes for a proposal
+   * @param _cId Claim ID
+   * @param _pId Proposal ID
+   */
+  function vote(bytes32 _cId, bytes32 _pId) external {
+    Claim storage c = claims[_cId];
+
+    require(c.status == ApplicationStatus.SUBMITTED, "SUBMITTED claim status required");
+    require(c.validators.has(msg.sender) == true, "Validator not in locked list");
+
+    Proposal storage p = c.proposalDetails[_pId];
+
+    require(p.from != address(0), "Proposal doesn't exists");
+
+    _voteFor(c, _pId);
+
+    if (p.votesFor.size() == c.n) {
+      c.chosenProposal = _pId;
+      calculateAndStoreAuditorRewards(c);
+
+      if (p.action == Action.APPROVE) {
+        changeSaleOrderStatus(c, ApplicationStatus.APPROVED);
+        validatorStakes.slash(p.validators, p.roles, p.fines);
+      } else {
+        changeSaleOrderStatus(c, ApplicationStatus.REJECTED);
+      }
+    }
+  }
+
+  function claimValidatorReward(bytes32 _cId) external {
+    Claim storage c = claims[_cId];
+
+    require(
+      c.status == ApplicationStatus.APPROVED || c.status == ApplicationStatus.REJECTED,
+      "Application status should be APPROVED or REJECTED");
+    require(c.validators.has(msg.sender) == true, "Validator not in locked list");
+    require(c.fees.validatorRewardPaidOut[msg.sender] == false);
+    validators.requireValidatorActiveWithAssignedActiveRole(msg.sender, CM_AUDITOR);
+
+    c.fees.validatorRewardPaidOut[msg.sender] = true;
+
+    if (c.fees.currency == Currency.ETH) {
+      msg.sender.transfer(c.fees.validatorReward);
+    } else if (c.fees.currency == Currency.GALT) {
+      galtToken.transfer(msg.sender, c.fees.validatorReward);
+    }
+  }
+
+  function claimGaltSpaceReward(bytes32 _cId) external {
+    require(msg.sender == galtSpaceRewardsAddress, "The method call allowed only for galtSpace address");
+
+    Claim storage c = claims[_cId];
+
+    /* solium-disable-next-line */
+    require(
+      c.status == ApplicationStatus.APPROVED || c.status == ApplicationStatus.REJECTED,
+      "Application status should be APPROVED or REJECTED");
+
+    require(c.fees.galtSpaceReward > 0, "Reward is 0");
+    require(c.fees.galtSpaceRewardPaidOut == false, "Reward is already paid out");
+
+    c.fees.galtSpaceRewardPaidOut = true;
+
+    if (c.fees.currency == Currency.ETH) {
+      msg.sender.transfer(c.fees.galtSpaceReward);
+    } else if (c.fees.currency == Currency.GALT) {
+      galtToken.transfer(msg.sender, c.fees.galtSpaceReward);
+    } else {
+      revert("Unknown currency");
+    }
+  }
+
+  function pushMessage(bytes32 _cId, string _text) external {
+    Claim storage c = claims[_cId];
+
+    require(c.status == ApplicationStatus.SUBMITTED, "SUBMITTED claim status required");
+    require(c.validators.has(msg.sender) == true || c.applicant == msg.sender, "Allowed only to an applicant or a validator");
+
+    uint256 id = c.messageCount;
+    c.messages[id] = Message(id, block.timestamp, msg.sender, _text);
```

Comment:
need an array of messages ids for frontend purposes

Reply:
we have just a counter `n`, so the range would be `0..n`